SwiftCompilerSources/Sources/Optimizer/Utilities/BorrowUtils.swift

```diff
@@ -0,0 +1,612 @@
+      self = .reborrow(Phi(arg)!)
+    default:
+      return nil
+    }
+  }
+  
+  var value: Value {
+    switch self {
+    case .beginBorrow(let bbi): return bbi
+    case .loadBorrow(let lbi): return lbi
+    case .partialApply(let pai): return pai
+    case .functionArgument(let arg): return arg
+    case .reborrow(let phi): return phi.value
+    }
+  }
+
+  init?(using operand: Operand) {
+    switch operand.instruction {
+    case is BeginBorrowInst, is LoadBorrowInst:
+      let inst = operand.instruction as! SingleValueInstruction
+      self = BeginBorrowValue(inst)!
+    case is BranchInst:
+      guard let phi = Phi(using: operand) else { return nil }
+      guard phi.isReborrow else { return nil }
+      self = .reborrow(phi)
+    default:
+      return nil
+    }
+  }
+
+  init?(resultOf borrowInstruction: BorrowingInstruction) {
+    if let value = borrowInstruction.instruction as? SingleValueInstruction,
+    let borrow = BeginBorrowValue(value) {
+      self = borrow
+    }
+    return nil
+  }
+
+  var hasLocalScope: Bool {
+    return !(value is FunctionArgument)
+  }
+
+  // non-nil if hasLocalScope is true
+  var baseOperand: Operand? {
+    switch value {
+    case let beginBorrow as BeginBorrowInst:
+      return beginBorrow.operand
+    case let loadBorrow as LoadBorrowInst:
+      return loadBorrow.operand
+    default:
+      return nil
+    }
+  }
+
+  public typealias ScopeEndingOperands =
+    LazyMapSequence<LazyFilterSequence<LazyMapSequence<UseList,
+                                                       EndBorrowOperand?>>,
+                                                       EndBorrowOperand>
+
+  /// The EndBorrows, reborrows (phis), and consumes (of closures)
+  /// that end the local borrow scope. Empty if hasLocalScope is false.
+  var scopeEndingOperands: ScopeEndingOperands {
+    value.uses.lazy.compactMap { EndBorrowOperand($0) }
+  }
+}
+
+/// Operand that uses a BeginBorrowValue, ending its borrow scope
+/// (and possibly propagating it to another borrow scope via
+/// reborrow).
+struct EndBorrowOperand {
+  enum Ownership { case endBorrow; case reborrow; case forwardingConsume;
+                   case destroyingConsume }
+  let ownership: Ownership
+  let operand: Operand
+
+  init?(_ operand: Operand) {
+    // This must handle every operand ownership that `endsLifetime`.
+    switch operand.ownership {
+    case .endBorrow: self.ownership = .endBorrow
+    case .reborrow: self.ownership = .reborrow
+    case .destroyingConsume: self.ownership = .destroyingConsume
+    case .forwardingConsume: self.ownership = .forwardingConsume
+    default: return nil
+    }
+    self.operand = operand
+  }
+
+  var instruction: Instruction { operand.instruction }
+}
+
+/// Find the enclosing borrow introducers for `value`. If `value` is
+/// owned, or introduces a borrow scope, then it is the single
+/// introducer for itself.
+func gather(borrowIntroducers: inout Stack<Value>, for value: Value,
```

Comment:
I changed it to `gatherBorrowIntroducers(for value: Value, in borrowIntroducers:`, although I don't understand the guidance. I wrote it the other way to try to fit what I thought was a better style.

I deliberately use `any Context` when it's being stored in a type that I want to avoid being generic or requiring specialization and in which the context is seldom accessed.